### PR TITLE
112 combining spatial transforms

### LINKED ISSUE #111: Port elastic deformation transform
a subtask of #62, port an elastic deformation transform into Monai

e.g.:
https://pypi.org/project/elasticdeform/

### LINKED ISSUE #112: implement stacked spatial transforms as coordinates ops + resampling
stacked spatial transforms such as #107 #108 #109 #110 #111 could be implemented as computing multiple operations on a grid and doing one step resampling

This ticket looks for a "vanilla" non-gpu implementation

https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.map_coordinates.html
https://pytorch.org/docs/stable/nn.functional.html#grid-sample

### PR DESCRIPTION
**Pull request by wyli**
Fixes #111 fixes #112.

### Description
implemented in pytorch for the combined spatial transforms:
- [x] affine transforms
- [x] randomised version of the transforms
- [x] random elastic deformations
- [x] unit tests
- [x] pipeline speed test (GPU/CPU)

This PR includes [a demo](https://github.com/Project-MONAI/MONAI/blob/2e1568a5751f65a42457d2a441d844cc4750922e/examples/transforms_demo_2d.ipynb) and [a speed test](https://github.com/Project-MONAI/MONAI/blob/2e1568a5751f65a42457d2a441d844cc4750922e/examples/transform_speed.ipynb).
Observations:

- the combined transforms are faster than the individual ones on both CPU/GPU
- can have a good preprocessing speed provided the output stays on GPU -- if these transforms are followed by some CPU operations, there're significant data transfer costs.
- need to run `torch.multiprocessing.set_start_method('spawn')` beforehand if both transforms and networks request GPU processing
- 3D elastic deformation field is approximated by gaussian smoothing over the coordinates because 3D cubic spline upsampling is not supported by pytorch

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/transforms/transforms.py [hunk 1]
```diff
@@ -206,10 +207,112 @@ def __call__(self, img):
...
+        if grid is None:
+            grid = create_grid(spatial_size)
+
+        spatial_dims = len(spatial_size)
+        affine = np.eye(spatial_dims + 1)
+        if self.rotate_params:
+            affine = affine @ create_rotate(spatial_dims, self.rotate_params)
+        if self.shear_params:
+            affine = affine @ create_shear(spatial_dims, self.shear_params)
+        if self.translate_params:
```
Comment (ericspod) [inline monai/transforms/transforms.py]: it is easier to fix the order of operations rather than provide a way of determining the order, but if we do fix the order I would say shift should be last. I think it would be expected that a transform would be applied to the image as it originally is located, if a shift precedes another operation this will distort the image relative to its new position.

File: monai/transforms/utils.py [hunk 2]
```diff
@@ -157,3 +156,108 @@ def one_hot(labels, num_classes):
...
+        if isinstance(radians, (list, tuple)):
+            radians = radians[0]
+        return np.array([[np.cos(radians), -np.sin(radians), 0.], [np.sin(radians), np.cos(radians), 0.], [0., 0., 1.]])
+
+    if spatial_dims == 3:
+        affine = None
+        if len(radians) >= 1:
+            affine = np.array([
+                [1., 0., 0., 0.],
+                [0., np.cos(radians[0]), -np.sin(radians[0]), 0.],
```
Comment (ericspod) [inline monai/transforms/utils.py]: The cos and sin calculations each appear twice, I don't think we can expect these to be optimized together so they should be pre-calculated before this line and referenced in the array.

File: monai/transforms/transforms.py [hunk 3]
```diff
@@ -206,10 +207,112 @@ def __call__(self, img):
...
+            return new_img.cpu().numpy()
+        return new_img
+
+
+class Affine:
+    """
+    transform ``img`` given the affine parameters.
+    """
+
+    def __init__(self,
```
Comment (ericspod) [inline monai/transforms/transforms.py]: Are the transform parameters randomly chosen at training time? I don't see how that is meant to happen quite yet.

Comment (wyli) [reply monai/transforms/transforms.py]: it's not in this PR yet, I'm working on it..

**Comment by wyli** [comment]
this is ready from my side, could you review? @ericspod @atbenmurray @Nic-Ma

File: tests/test_gaussian_filter.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+import unittest
+
+import numpy as np
+import torch
+
+from monai.networks.layers.simplelayers import GaussianFilter
+
+
+class IntensityNormTestCase(unittest.TestCase):
```
Comment (Nic-Ma) [inline tests/test_gaussian_filter.py]: Typo, should not be "IntensityNorm".

File: monai/transforms/transforms.py [hunk 5]
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
...
+        if self.scale_range:
+            self.scale_params = [self.R.uniform(-f, f) + 1.0 for f in self.scale_range if f is not None]
+
+    def __call__(self, spatial_size=None, grid=None):
+        """
+        Returns:
+            a 2D (3xHxW) or 3D (4xHxWxD) grid.
+        """
+        self.randomize()
+        _affine_grid = AffineGrid(self.rotate_params, self.shear_params, self.translate_params, self.scale_params,
```
Comment (Nic-Ma) [inline monai/transforms/transforms.py]: I think we should not use "_" for local variables. It's used for protected or private member variables.

File: monai/transforms/transforms.py [hunk 6]
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
...
+        grid = create_grid(spatial_size)
+        if self.do_transform:
+            grid = torch.tensor(grid).to(self.device)
+            gaussian = GaussianFilter(3, self.sigma, 3., device=self.device)
+            grid[:3] += gaussian(self.rand_offset[None])[0] * self.magnitude
+            grid = self.rand_affine_grid(grid=grid)
+        return self.resampler(img, grid, mode)
+
+
+if __name__ == "__main__":
```
Comment (Nic-Ma) [inline monai/transforms/transforms.py]: Suggest to remove test code from official release code.

File: monai/transforms/utils.py [hunk 7]
```diff
@@ -208,3 +209,136 @@ def generate_pos_neg_label_crop_centers(label, size, num_samples, pos_ratio, ran
...
+        return np.array([
+            [1., coefs[0], coefs[1], 0.],
+            [coefs[2], 1., coefs[3], 0.],
+            [coefs[4], coefs[5], 1., 0.],
+            [0., 0., 0., 1.],
+        ])
+    raise NotImplementedError
+
+
+def create_scale(spatial_dims, s):
```
Comment (Nic-Ma) [inline monai/transforms/utils.py]: Why not use unified argument name "coefs" for all factor matrix?
I think "s" and "t" are too simple and unclear.
Thanks.

File: tests/test_random_affine.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+             translate_range=[2, 1],
+             scale_range=[.1, .2],
+             as_tensor_output=True,
+             device=None), {'img': torch.arange(64).reshape((1, 8, 8)), 'spatial_size': (3, 3)},
+        torch.tensor([[[27.3614, 18.0237, 8.6860], [40.0440, 30.7063, 21.3686], [52.7266, 43.3889, 34.0512]]])
+    ],
+]
+
+
+class TestRandAffineGrid(unittest.TestCase):
```
Comment (Nic-Ma) [inline tests/test_random_affine.py]: Typo, should be "TestRandAffine" instead of "TestRandAffineGrid"?

File: tests/test_affine.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        {'img': np.arange(8).reshape((1, 2, 2, 2)), 'spatial_size': (4, 4, 4)},
+        np.array([[[[0., 0., 0., 0.], [0., 0., 0., 0.], [0., 0., 0., 0.], [0., 0., 0., 0.]],
+                   [[0., 0., 0., 0.], [0., 0.25, 0., 0.], [0., 0.375, 0.125, 0.], [0., 0., 0., 0.]],
+                   [[0., 0., 0., 0.], [0., 0.75, 0.5, 0.], [0., 0.875, 0.625, 0.], [0., 0., 0., 0.]],
+                   [[0., 0., 0., 0.], [0., 0., 0., 0.], [0., 0., 0., 0.], [0., 0., 0., 0.]]]])
+    ],
+]
+
+
+class TestResample(unittest.TestCase):
```
Comment (Nic-Ma) [inline tests/test_affine.py]: Typo, should be "TestAffine" instead of "TestResample"?

File: tests/test_random_elastic_3d.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            'magnitude_range': (.3, .3), 'sigma_range': (1., 2.), 'prob': 0.9, 'rotate_range': [1, 1, 1],
+            'as_tensor_output': False, 'device': None, 'spatial_size': (2, 2, 2)
+        },
+        {'img': torch.arange(27).reshape((1, 3, 3, 3)), 'mode': 'bilinear'},
+        np.array([[[[6.016205, 2.3112855], [12.412318, 11.182229]], [[14.619441, 6.9230556], [17.23721, 16.506298]]]]),
+    ],
+]
+
+
+class TestRand2DElastic(unittest.TestCase):
```
Comment (Nic-Ma) [inline tests/test_random_elastic_3d.py]: Typo, should be "TestRand3DElastic".

File: monai/transforms/transforms.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/transforms/transforms.py]: Maybe you can support user's definition of different params in order, and construct a "Sub transform chain" based on the params order.
Thanks.

File: examples/transforms_demo_2d.ipynb [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# 2D image transformation demo\n",
+    "\n",
+    "This demo shows how to apply 2D transforms in MONAI.\n",
+    "Main features:\n",
+    "    - Random elastic transforms implemented in native Pytorch\n",
+    "    - Easy-to-use interfaces that are designed and implmented in the pythonic way\n",
```
Comment (Nic-Ma) [inline examples/transforms_demo_2d.ipynb]: Typo, should be "implemented" instead of "implmented"?

File: examples/transforms_demo_2d.ipynb [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+    "    \n",
+    "Find out more in MONAI's wiki page: https://github.com/Project-MONAI/MONAI/wiki"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Before running this demo\n",
+    "**please download the GLAS (gland segmentation in histology images) challenge data from\n",
```
Comment (Nic-Ma) [inline examples/transforms_demo_2d.ipynb]: Suggest to add ":" and "\n" at end of this line, then the notebook style will be normal.

File: examples/transform_speed.ipynb [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+    "    ToTensor()\n",
+    "])    \n",
+    "\n",
+    "segtrans = Compose([\n",
+    "    AddChannel(),\n",
+    "    ToTensor()\n",
+    "])    \n",
+    "    \n",
+    "ds = NiftiDataset(images, segs, transform=imtrans, seg_transform=segtrans)\n",
+    "loader = DataLoader(ds, batch_size=3, num_workers=8)\n",
```
Comment (Nic-Ma) [inline examples/transform_speed.ipynb]: Why num_workers is 8 instead of 4 here? It's different from your other CPU test.
Thanks.

File: examples/transform_speed.ipynb [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+    "\n",
+    "segtrans = Compose([\n",
+    "    AddChannel(),\n",
+    "    Rotate(angle=45.),\n",
+    "    UniformRandomPatch((64, 64, 64)),\n",
+    "    ToTensor()\n",
+    "])    \n",
+    "    \n",
+    "ds = NiftiDataset(images, segs, transform=imtrans, seg_transform=segtrans)\n",
+    "loader = DataLoader(ds, batch_size=3, num_workers=4, pin_memory=torch.cuda.is_available())\n",
```
Comment (Nic-Ma) [inline examples/transform_speed.ipynb]: As num_workers obviously affect CPU performance and seed, is this "num_workers=4" the best value on your CPU?
And maybe you need to state your CPU type in the markdown introduction.
Thanks.

File: monai/transforms/transforms.py [hunk 15]
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
...
+            spatial_size (2 ints): specifying output image spatial size [h, w, d].
+            mode ('nearest'|'bilinear'): interpolation order. Defaults to 'self.mode'.
+        """
+        spatial_size = spatial_size or self.spatial_size
+        mode = mode or self.mode
+        self.randomize(spatial_size)
+        grid = create_grid(spatial_size)
+        if self.do_transform:
+            grid = torch.tensor(grid).to(self.device)
+            gaussian = GaussianFilter(3, self.sigma, 3., device=self.device)
```
Comment (Nic-Ma) [inline monai/transforms/transforms.py]: As GaussianFilter contains "torch.nn.Parameter", should add "with torch.no_grad():" when run this operation?
Thanks.

File: monai/transforms/transforms.py [hunk 16]
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
...
+                 device=None):
+        self.rotate_params = rotate_params
+        self.shear_params = shear_params
+        self.translate_params = translate_params
+        self.scale_params = scale_params
+
+        self.as_tensor_output = as_tensor_output
+        self.device = device
+
+    def __call__(self, spatial_size=None, grid=None):
```
Comment (Nic-Ma) [inline monai/transforms/transforms.py]: If this class is to **generate grid** instead of **transform image**, why not define it as a utility?

File: monai/transforms/transforms.py [hunk 17]
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
...
+        """
+        self.affine_grid = AffineGrid(rotate_params,
+                                      shear_params,
+                                      translate_params,
+                                      scale_params,
+                                      as_tensor_output=True,
+                                      device=device)
+        self.resampler = Resample(padding_mode, as_tensor_output=as_tensor_output, device=device)
+
+    def __call__(self, img, spatial_size, mode='bilinear'):
```
Comment (Nic-Ma) [inline monai/transforms/transforms.py]: Can we use the same **spatial_size** and **mode** for all images in training?
If yes, strongly suggest to move them to __init__(), then we can compose this transform with others as a transform chain.
Otherwise, I feel it's not easy to use it with other transforms together.
Thanks.

File: monai/transforms/transforms.py (hunk 15 above)
Comment (wyli) [reply monai/transforms/transforms.py]: i set requires_grad to False https://github.com/Project-MONAI/MONAI/pull/131/files/871b074df3fada02a2d24507fe2a5ca81cddee8a?file-filters%5B%5D=.py#diff-94f123b1e6105a0be041ce6ef8461918R46

File: monai/transforms/transforms.py (hunk 16 above)
Comment (wyli) [reply monai/transforms/transforms.py]: i think the user might load "grid" data directly and apply grid transform here

**Comment by wyli** [comment]
> Added several comments inline.
> As this PR is very big, need to double-check the math computations carefully later.
> Thanks.

thanks! addressed most of the comments, the pr is too large to include ordering of the operations

File: monai/transforms/transforms.py (hunk 15 above)
Comment (Nic-Ma) [reply monai/transforms/transforms.py]: OK, sounds good.

File: monai/transforms/transforms.py (hunk 16 above)
Comment (Nic-Ma) [reply monai/transforms/transforms.py]: OK.
1. It's acceptable.
2. But I am not sure whether we should explicitly define "transforms" only for images.
@ericspod @atbenmurray , do you guys have any comments?
Thanks.

Comment (wyli) [reply monai/transforms/transforms.py]: perhaps that's not needed, e.g. we allow the users to apply "seg_transforms" on classification labels as well https://github.com/Project-MONAI/MONAI/blob/master/monai/data/nifti_reader.py#L113, I think it's a nice feature for flexibilities

File: monai/networks/layers/simplelayers.py [hunk 18]
```diff
@@ -11,6 +11,9 @@
 
 import torch
 import torch.nn as nn
+import torch.nn.functional as nnf
```
Comment (Nic-Ma) [inline monai/networks/layers/simplelayers.py]: Suggest to use "F" or "f" according to almost all PyTorch official examples.
